File mechanism wants to be merged to main

KristianIvarsson opened the pull request:
The basic functionality is ready and done

Probably some CLI features will be added later on

lazan commented on middleware/file/documentation/file.development.md:
````diff
@@ -0,0 +1,75 @@
+   std::filesystem::remove( path);
+...
+```
+
````
Shouldn't we include tx_begin(), tx_commit() in these examples? Just to make it clear to the reader that transaction is the major part of the whole file-manager...

KristianIvarsson replied:
Good idea, we'll put that into some commit sometime

lazan commented on middleware/file/include/file/api/c/file.h:
```diff
@@ -0,0 +1,39 @@
+/* releases the memory allocated during acquire or attempt */
+extern void casual_file_release( char* path);
+
+#ifdef __cplusplus
```
Pretty good. I think this is more than good enough. 
I would argue that a _flag_ to `casual_file_reserve` would be a more _posix style_ api. And, more importantly, opens up future non intrusive features.
But I'm fine as is.

KristianIvarsson replied:
The documentation (description) is wrong though

lazan commented on middleware/file/include/file/code.h:
Nice

lazan commented on middleware/file/include/file/factory.h:
```diff
@@ -0,0 +1,57 @@
+      {
+         namespace blocking
+         {
+            inline auto request( std::filesystem::path path)
```
Why are these inline? Any particular reason?

KristianIvarsson replied:
Hmm, it must be some leftovers to avoid ODR in some earlier design and I guess it can be removed now

lazan commented on middleware/file/include/file/manager/admin/model.h:
```diff
@@ -0,0 +1,53 @@
+               CASUAL_SERIALIZE( working);
+               CASUAL_SERIALIZE( pending);
+            )
+         };
```
Simple and clean.

lazan commented on middleware/file/include/file/manager/handle.h:
```diff
@@ -0,0 +1,26 @@
+      using dispatch_type = decltype( common::message::dispatch::handler( common::communication::ipc::inbound::device()));
+
+      //! @returns all the handlers for service manager
+      dispatch_type create( State& state);
```
_file-manager_

lazan commented on middleware/file/source/api/c/file.cpp:
```diff
@@ -0,0 +1,129 @@
+      } // local
+
+
+      extern "C" 
```
I don't think you need `extern C` in the TU

lazan commented on middleware/file/source/manager/resource.cpp:
I strongly prefer not to send anything from _resouce.cpp_. If possible all send should be in _handle.cpp_ . It's more symmetric.  receive-request -> send-reply at in the same function. 
This would mean that you return the message that _handle_ should send back to caller. Then it's easy to unittest the resource as a bonus.

KristianIvarsson replied:
Maybe agreed, but replies are not always gonna be sent but instead sent in other scenarios, such as when a _commit_ is issued, maybe replies for _reserve_ is gonna be sent (as well)

Maybe I don't understand what you're aiming for ?

lazan commented on middleware/file/source/manager/handle.cpp:
```diff
@@ -0,0 +1,143 @@
+                     {
+                        common::Trace trace{ "file::manager::handle::local::transaction::commit::request"};
+
+                        resource::commit( state, message);
```
I would like `resource::commit(..)` to return the message and this function sends the reply to caller. Much clearer that the reply is sent.

lazan reviewed (changes requested):
First pass :)

lazan commented on middleware/file/include/file/factory.h:
This is a _message factory_? 
* should not the file name reflect that?
* why header only?

In the rest of `causal` I think _factory_ is used for dynamic dispatch factories (sort of classic factory). We have very few. 
This is more of a _create helper_? 
`file/message/create.h` ?

KristianIvarsson replied:
Give a name suggestion

It is used from 2 binaries and thus we need to build it as a object file or an archve and link it into those two binaries so this was just simpler to have it as header only

lazan replied:
I just gave a name suggestion ..?
Yeah, either an archive or you just keep the object-file(s) in a python-list and add them in the binaries that needs them.

lazan replied:
`file::message::create::[non::]blocking::request( ...)`

KristianIvarsson replied:
Did some more proper renaming of namespaces (and removed a forgotten ìnline` directive, but didn't change any file structure)

lazan commented on middleware/file/source/manager/handle.cpp:
```diff
@@ -0,0 +1,143 @@
+                        
+                        auto reply = common::message::reverse::type( message);
+                        state.multiplex.send( message.process.ipc, reply);
+                     };
```
Have a `resource::prepare` for symmetry.

KristianIvarsson replied:
Greate point (fixed now)

KristianIvarsson committed:
file: introduce file reserve mechanism within xa transactions

A reserved file is isolated from other transactions as long as all
participants interacts through this API

References issue #13

lazan reviewed (approved):
Nice.
